Upgrade ESLint to v8

willdurand opened the pull request:
Fixes https://github.com/mozilla/addons-linter/issues/3985
Fixes https://github.com/mozilla/addons-linter/issues/3062
Fixes https://github.com/mozilla/addons-linter/issues/3741
Fixes https://github.com/mozilla/addons-linter/issues/3960

---

This PR upgrades `eslint` to v8, which will allow us to support the 2022 syntax.

## TODO

- [x] update `addons-scanner-utils` when it supports ESLint v8
- [x] update `eslint-config-amo` when it supports ESLint v8
- [x] add support for ESLint v8 in https://github.com/mozilla/eslint-plugin-no-unsanitized

willdurand commented on tests/unit/scanners/test.html.js:
```diff
@@ -102,6 +102,7 @@ describe('HTML', () => {
     expect(linterMessages.length).toEqual(0);
   });
 
+  // eslint-disable-next-line jest/expect-expect
```
done

willdurand commented on package.json:
```diff
@@ -87,15 +88,13 @@
-    "eslint-config-amo": "4.11.0",
+    "eslint-config-amo": "5.1.0",
     "eslint-plugin-async-await": "0.0.0",
-    "eslint-plugin-promise": "5.1.1",
```
not used it seems?

willdurand commented on package.json:
```diff
@@ -87,15 +88,13 @@
     "doctoc": "2.1.0",
-    "eslint-config-amo": "4.11.0",
+    "eslint-config-amo": "5.1.0",
     "eslint-plugin-async-await": "0.0.0",
```
I am questioning the usefulness of this plugin: I don't think it is used in most of the other projects and we do use async/await everywhere now. I would like to remove this dev dependency: WDYT?

rpl replied:
@willdurand it looks that it does only provides two "code formatting" rules, which I would guess that prettier may also be taking care of. Also the package version is v0.0.0 and last release 5 years ago, definitely :+1:  on my side on getting rid of it.

willdurand commented:
The last commit about removing `eslint-plugin-promise` shows that pinning the `eslint-plugin-no-unsanitized` version (using a git commit) fixes the issue for this project (i.e. `eslint-plugin-no-unsanitized` with ESLint v8 support works as intended here).

Removing `eslint-plugin-promise` was "needed" because [this plugin does not support ESLint v8 yet](https://github.com/xjamundx/eslint-plugin-promise/issues/218) and `npm` fails in some cases because of the peer dependency mismatch. That's annoying because we are blocked by this dev dependency now.

Apart from that, I'd like to remove `eslint-plugin-async-await`, which I don't find particularly useful. Thoughts?

willdurand commented:
re-requesting a review not for the review itself but to make sure you saw the last comment.

willdurand commented:
I switched to v4.0.0 of the `eslint-plugin-no-unsanitized` plugin.

willdurand commented:
@rpl for when you have time, this is ready for a review with two caveats:

- `eslint-plugin-promise` removed because it is currently not compatible
- `eslint-plugin-async-await` isn't removed yet but I would like to remove it, see: https://github.com/mozilla/addons-linter/pull/3984#discussion_r740788256

I am not sure when `eslint-plugin-promise` will get proper support for ESLint v8 so I am thinking about removing it here now and adding it back in a different patch when it is compatible? That would allow us to move forward given that `eslint-plugin-promise` is a dev dependency (and not a lib we use in the core of the linter).

rpl commented:
> @rpl for when you have time, this is ready for a review with two caveats:
> 
>     * `eslint-plugin-promise` removed because it is currently not compatible
> 
>     * `eslint-plugin-async-await` isn't removed yet but I would like to remove it, see: [Upgrade ESLint to v8 #3984 (comment)](https://github.com/mozilla/addons-linter/pull/3984#discussion_r740788256)
> 
> 
> I am not sure when `eslint-plugin-promise` will get proper support for ESLint v8 so I am thinking about removing it here now and adding it back in a different patch when it is compatible? That would allow us to move forward given that `eslint-plugin-promise` is a dev dependency (and not a lib we use in the core of the linter).

100% agree on `eslint-plugin-async-await` (https://github.com/mozilla/addons-linter/pull/3984#discussion_r741861581).

For `eslint-plugin-promise` instead, it would be nice to bring it back as soon as we can, I'm ok to remove it to unblock us to upgrade to eslint v8 along with filing a follow up issue to bring it back.

rpl commented on package.json:
```diff
@@ -43,9 +43,10 @@
   },
   "homepage": "https://github.com/mozilla/addons-linter#readme",
   "dependencies": {
+    "@babel/eslint-parser": "7.15.8",
```
It looks that "babel-eslint" used to be a dev dependency, while "@babel/eslint-parser" is listed as a direct dependency, 

Based on the content of this PR it looks that @babel/eslint-parser it is still only used while running eslint on the addons-linter sources for development purpose, and it may still not be needed as a strict dependency for an addons-linter user.

is there something else that I may be missing to notice?

willdurand replied:
No, that was a mistake. Good catch, thanks.

rpl reviewed (changes requested):
@willdurand lgtm, just one detail related to @babel/eslint-parser being listed as a dependency instead of a dev dependency. Everything else looks good to me as is.

willdurand commented:
I moved the babel parser to the list of dev dependencies and removed the async-await eslint plugin. I also filed an issue for the eslint-plugin-promise: https://github.com/mozilla/addons-linter/issues/4016

rpl reviewed (approved):
Thanks!
